File: tests/pyre.lib/timers/process_timer_example.cc
// -*- c++ -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2023 all rights reserved


// get the timers
#include <pyre/timers.h>
#include <pyre/journal.h>
// support
#include <thread>
#include <cassert>
// access the {chrono} literals
using namespace std::literals;


// type aliases
using proctimer_t = pyre::timers::process_timer_t;


// verify that two timers that have the same name share the same movement
int main() {
    // make a timer
    proctimer_t t("tests.timer");
    // start it
    t.start();

    // do something
    double sum=0;
    for (int i=0; i < 1000*1000; ++i) {
        sum += i;
    }

    // stop the timer
    t.stop();

    // make a channel
    pyre::journal::debug_t channel("tests.timer");
    // activate it
    // channel.activate();
    // show me
    channel
        // show me the sum
        << "sum: " << sum << pyre::journal::newline
        // show me the elapsed time
        << "elapsed time: " << t.us()
        << pyre::journal::endl(__HERE__);

    // all done
    return 0;
}


// end of file

File: tests/pyre.ext/timers/process_timer_example.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2023 all rights reserved


def test():
    """
    Exercise the common use case with a cpu timer
    """
    # access the timer bindings
    from pyre.extensions.pyre.timers import ProcessTimer

    # make a timer
    t = ProcessTimer(name="tests.timer")
    # start it
    t.start()

    # do something
    s = sum(range(10**6))

    # stop the timer
    t.stop()

    # get the journal
    import journal
    # make a channel
    channel = journal.debug(name="pyre.timers")
    # activate it
    # channel.activate()
    # log the elapsed time
    channel.log(f"elapsed: {t.ms()}ms")

    # all done
    return


# main
if __name__ == "__main__":
    # run the test
    test()


# end of file
